File: go/day_2/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func main() {
	partOne()
	partTwo()
}

func partOne() {
	file, err := os.Open("passwords.txt")
	check(err)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	correctCount := 0

	for scanner.Scan() {
		fields := strings.Fields(scanner.Text())

		minmax := strings.Split(fields[0], "-")
		min, err := strconv.Atoi(minmax[0])
		check(err)
		max, err := strconv.Atoi(minmax[1])
		check(err)

		letter := rune(fields[1][0])

		letterCount := 0

		for _, v := range fields[2] {
			if letter == v {
				letterCount++
			}
		}

		if letterCount >= min && letterCount <= max {
			correctCount++
		}
	}

	fmt.Println("Part One:", correctCount)
	return
}

func partTwo() {
	file, err := os.Open("passwords.txt")
	check(err)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	correctCount := 0

	for scanner.Scan() {
		fields := strings.Fields(scanner.Text())

		position := strings.Split(fields[0], "-")
		position1, err := strconv.Atoi(position[0])
		check(err)
		position2, err := strconv.Atoi(position[1])
		check(err)

		letter := rune(fields[1][0])
		letter1 := rune(fields[2][position1-1])
		letter2 := rune(fields[2][position2-1])

		if letter1 == letter && letter2 != letter {
			correctCount++
		} else if letter1 != letter && letter2 == letter {
			correctCount++
		}
	}

	fmt.Println("Part Two:", correctCount)
	return
}

func check(err error) {
	if err != nil {
		log.Fatal(err)
	}
}

File: rust/day_2/src/main.rs
use std::fs::File;
use std::io::{prelude::*, BufReader};

fn main() {
    part_one();
    part_two();
}

fn part_one() {
    let file = File::open("passwords.txt").expect("");
    let reader = BufReader::new(file);

    let mut correct_count: u32 = 0;

    for line in reader.lines() {
        let line = line.expect("");
        let mut line = line.split_whitespace();

        let mut min_max = line.next().expect("").split("-");

        let min: u32 = min_max.next().expect("").parse().expect("");
        let max: u32 = min_max.next().expect("").parse().expect("");

        let expected_letter = line.next().expect("").chars().next().expect("");

        let mut letter_count: u32 = 0;

        for letter in line.next().expect("").chars() {
            if letter == expected_letter {
                letter_count = letter_count + 1;
            }
        }

        if (letter_count >= min) && (letter_count <= max) {
            correct_count = correct_count + 1;
        }
    }

    println!("Part One: {}", correct_count)
}

fn part_two() {
    let file = File::open("passwords.txt").expect("");
    let reader = BufReader::new(file);

    let mut correct_count: u32 = 0;

    for line in reader.lines() {
        let line = line.expect("");
        let mut line = line.split_whitespace();

        let mut min_max = line.next().expect("").split("-");

        let position1: usize = min_max.next().expect("").parse().expect("");
        let position2: usize = min_max.next().expect("").parse().expect("");

        let letter = line.next().expect("").chars().next().expect("");

        let mut letters = line.next().expect("").chars();

        let letter1 = letters.nth(position1 - 1).expect("");
        let letter2 = letters.nth(position2 - position1 - 1).expect("");

        if (letter1 == letter) && (letter2 != letter) {
            correct_count = correct_count + 1;
        } 
        else if (letter2 == letter) && (letter1 != letter) {
            correct_count = correct_count + 1;
        }
    }

    println!("Part Two: {}", correct_count)
}